### PR TITLE
Documentation/rest api web

### PR DESCRIPTION
**Pull request by dmoisej**
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see [Coding Guidelines](/doc/CODING.md))
- [ ] meta data is updated (e.g. README.md of plugins and [METADATA.ini](/doc/METADATA.md))

## Review

Reviewers will usually check the following:

- [ ] Code is conforming to [our Coding Guidelines](/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](/doc/DESGIN.md)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

## Labels

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the build server is happy** and also you
  say that everything is ready to be merged.

@markus2330 @sanssecours please review my pull request

### CODE REVIEW

**Comment by dmoisej** [comment]
@markus2330 @sanssecours the pull request is related to the https://github.com/ElektraInitiative/libelektra/pull/2746

It is just another copy as I was forced to refork the repository. The reason for that was, that I mistakenly  added commits to the master, which are not merged into this (main) repository, so I get those changes on every pull request.

Deleting commits with the git rebases didn't help me. Additionally I had some problems with jenkins tool, which checks, if the commit can be merged or not.

File: src/tools/web/README.md [hunk 1]
```diff
@@ -114,6 +114,53 @@ Elektra web consists of multiple components:
...
+## Test REST API on localhost
+
+In order to test API on localhost, your have to start elektrad instance. You can do it in two ways:
+
+- run manually
+
+  - `cd libelektra/src/tools/web`
+  - `cd elektrad`
+  - `npm install`
+  - `npm start` (replaces `kdb run-elektrad`)
```
Comment (markus2330) [inline src/tools/web/README.md]: What means "replaces" here? From source you use `npm start`, if installed `kdb run-elektrad`. It seems to be that this instructions duplicate the instructions above or web/elektrad?

File: src/tools/web/README.md [hunk 2]
```diff
@@ -114,6 +114,53 @@ Elektra web consists of multiple components:
...
+
+- by installing elektrad tool together with Elektra and run it
+  - please see the documentation [here](https://www.libelektra.org/tools/web) on how to install and run the elektra-web tool
```
Comment (markus2330) [inline src/tools/web/README.md]: Is this self-reference on purpose?

File: src/tools/web/README.md [hunk 3]
```diff
@@ -114,6 +114,53 @@ Elektra web consists of multiple components:
...
+  
+Now the server is runing on [http://localhost:33333](http://localhost:33333). After that you can test API with help of Postman or other tool, which allows to send REST API requests.
+
+As an example:
+
+Let's create the new key-value pair `user/test`:
+
+```sh
+kdb set user/test 5
```
Comment (markus2330) [inline src/tools/web/README.md]: How to set via REST API?

File: src/tools/web/README.md [hunk 4]
```diff
@@ -114,6 +114,53 @@ Elektra web consists of multiple components:
...
+    "exists": true,
+    "name": "test",
+    "path": "user/test",
+    "ls": [
+        "user/test"
+    ],
+    "value": "5",
+    "meta": ""
+}
+```
```
Comment (markus2330) [inline src/tools/web/README.md]: Please refer to the API docu for this.

File: src/tools/web/README.md (hunk 1 above)
Comment (dmoisej) [reply src/tools/web/README.md]: Actually, I just copy pasted what is written in https://www.libelektra.org/tools/web in the section "Running from source".

It means, that all 4 commands:
- `cd libelektra/src/tools/web`
- `cd elektrad`
- `npm install`
- `npm start`

replace kdb run-elektrad

File: src/tools/web/README.md (hunk 2 above)
Comment (dmoisej) [reply src/tools/web/README.md]: oh sorry, I didn't notice that, I will improve this

File: src/tools/web/README.md (hunk 4 above)
Comment (dmoisej) [reply src/tools/web/README.md]: Will improve

**Comment by dmoisej** [comment]
@markus2330, I updated the guide, please check

File: src/tools/web/README.md (hunk 1 above)
Comment (markus2330) [reply src/tools/web/README.md]: > Actually, I just copy pasted

Please don't do this.

https://www.libelektra.org/tools/web is generated from this document.

**Comment by markus2330** [comment]
The installation procedure is still copy&pasted?

**Comment by dmoisej** [comment]
@markus2330 , no, the installation of the web-elektrad tool is not copy pasted. Right now, I just added the note, "please see the section Building with elektra-web Tool". 

The only thing, which is copy pasted, is four commands to start the server manually:
cd libelektra/src/tools/web
cd elektrad
npm install
npm start

But everything others is written newly.

**Comment by sanssecours** [comment]
> It is just another copy as I was forced to refork the repository. The reason for that was, that I mistakenly added commits to the master, which are not merged into this (main) repository, so I get those changes on every pull request.

You can always just create another branch based on the `master` branch (or any other branch) of this repository. Usually you do not need to

- create a new fork of a repository you already cloned, or 
- create a new PR (just rebase and force-push your changes)

. Anyway, I would recommend you read the book [“Git Pro”](https://git-scm.com/book/en/v2) to learn a little bit more about the versioning tool. I also recommend you use a GUI, which usually simplifies working with `git` quite a lot. If you have a Mac (or use Windows), then I recommend you use [Fork](https://git-fork.com). [Tower](http://git-tower.com) is even better in my experience, but is quite expensive and uses a really user-unfriendly licensing model.

**Comment by dmoisej** [comment]
@sanssecours, thank you for your recommendation. I didn't know, that I could create the new branch from master of the main repository as I haven't worked with forked repositories before.

**Comment by dmoisej** [comment]
@sanssecours I will try to fix code formatting later today.

**Comment by dmoisej** [comment]
@sanssecours , @markus2330  I was forced to change the output format from json to sh because in the doc/CODING.md, on the line 376, it is specified:

- `README.md` and tutorials should be written exclusively with shell recorder syntax so that we know that the code in the tutorial produces output as expected

I checked then README.md with prettier tool after the changes and everything works well right now.

Additionally to this, I tried to reformat the README.md file together with scripts/reformat-markdown script from the libelektra project root. 
After running:

```sh scripts/reformat-markdown src/tools/web/README.md``` 

I've got the error:

```scripts/reformat-markdown: Zeile 7: $'\r': Kommando nicht gefunden.```

**Comment by sanssecours** [comment]
> @sanssecours , @markus2330 I was forced to change the output format from json to sh because in the doc/CODING.md, on the line 376, it is specified:
> 
> * `README.md` and tutorials should be written exclusively with shell recorder syntax so that we know that the code in the tutorial produces output as expected

Please do not use incorrect language specifiers. The code block in question contains JSON code, not code for the Bourne shell (aka `sh`). The language specifier should reflect this.

You do **not need** to write this tutorial using [(Markdown) Shell Recorder syntax](https://master.libelektra.org/tests/shell/shell_recorder/tutorial_wrapper), even if the sentence you quoted might suggest otherwise.

> After running:
> 
> `sh scripts/reformat-markdown src/tools/web/README.md`
> 
> I've got the error:
> 
> `scripts/reformat-markdown: Zeile 7: $'\r': Kommando nicht gefunden.`

Interesting. Did you maybe [change the line endings](https://stackoverflow.com/questions/11616835/r-command-not-found-bashrc-bash-profile) of `reformat-markdown` from Unix (`\n`) to DOS line endings (`\r\n`)?

**Comment by dmoisej** [comment]
@sanssecours , 

> Please do not use incorrect language specifiers. The code block in question contains JSON code, not code for the Bourne shell (aka `sh`). The language specifier should reflect this.
> 
> You do **not need** to write this tutorial using [(Markdown) Shell Recorder syntax](https://master.libelektra.org/tests/shell/shell_recorder/tutorial_wrapper), even if the sentence you quoted might suggest otherwise.

Okay but if the build is failed because I specify any other language, then sh, so what should I do right now? Should I change something? If yes, than what exactly should I change and on what? Or will my pull request will be merged?

> Interesting. Did you maybe [change the line endings](https://stackoverflow.com/questions/11616835/r-command-not-found-bashrc-bash-profile) of `reformat-markdown` from Unix (`\n`) to DOS line endings (`\r\n`)?

I tried it quickly form my other laptop, where I have windows installed. In the evening I will try to execute this script on other laptop, where I have linux installed.

**Comment by sanssecours** [comment]
> If yes, than what exactly should I change and on what?

I just pushed a [commit](https://github.com/ElektraInitiative/libelektra/pull/2770/commits/06cae4567f2a3427d69c8174d4dfc38ea80fb51e) that contains my suggested changes.

> Or will my pull request will be merged?

I think as soon as Markus and the build server are happy with the status of the PR, Markus will merge your updates.

**Comment by dmoisej** [comment]
@sanssecours thank you very much for help.

File: src/tools/web/README.md (hunk 4 above)
Comment (markus2330) [reply src/tools/web/README.md]: Not yet done?

Comment (dmoisej) [reply src/tools/web/README.md]: @markus2330 it is done. In the file I showed, how can you set a value via curl

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**